### DISCUSSION TITLE
Possibility of having a GAEN-filter

### DISCUSSION THREAD

**Discussion by Werner-G** [theengs/gateway#22] [Ideas]
GAEN (https://en.wikipedia.org/wiki/Exposure_Notification) is a worldwide used protocol used for COVID-19-monitoring with your smartphone. As a result, lots of messages with always new id's coming from your phone spamming any BLE-scanner, also this gateway. I wonder if it is possible to set up a filter to ignore this messages before transforming them into MQTT.

**Comment by koenvervloesem** [comment]
Maybe this could be generalized to a filter in `theengsgw.conf`, like this:

```json
{
    "filter_decoders": ["IBEACON", "GAEN", "MS-CDP"]
}
```

So you can specify exactly which types of devices you're not interested in.

**Reply by Werner-G** [reply]
Great idea!

**Reply by 1technophile** [reply]
Good idea, We could have the same principle of white-list/black-list that we use with OMG for MAC Address but for MODEL_ID (and align both project in terms of functionalities). So that at the end OMG can white list/black list Mac address and MODEL_ID through an MQTT command and also Theengs but through the configuration file.

**Reply by Werner-G** [reply]
I'm not so familiar with guthub. Do I have to issue a 'pull request' to push this idea?

**Comment by drhirn** [comment]
And why not use the same black/whitelists with MAC as OMG does?

**Reply by 1technophile** [reply]
We could do that indeed, ideally we should have model_id, white-list and black-list

**Reply by drhirn** [reply]
Would be very happy to see that. Unfortunately I can't help you with the coding. But with testing.

**Comment by echavet** [comment]
The ble_gateway does publish a lot of MQTT messages.
I think it can save CPU and network to filter some of them. I don't know the side effects but in my case, I still get the messages from my thermometers!!

I've just added the "if" condition before the last line of the file ble_gateway.py so as to use the discovery_filter parameter to filter. It is far less verbose. I will check for side effects...

```
if (not pub_device["model_id"] in self.discovery_filter):            
            self.publish(device_data, self.pub_topic + "/" + pub_device_uuid)
```

**Reply by 1technophile** [reply]
Agreed, extending the scope of the discovery filter to all the messages make sense

**Comment by MrSco** [comment]
was this ever implemented? I'd love to only send mqtt messages for the handful of MAC addresses I care about instead of 150+ mqtt messages as it is now

**Reply by DigiH** [reply]
@MrSco what other messages are you getting, and what kind of Theengs Gateway installation are you using - pip, HA Add-on, Docker …?

**Reply by MrSco** [reply]
its finding all bluetooth hardware mac addresses... I want it to only find a list of mac addresses i give it. i believe bluez/bleak has a filter by mac address feature. theengs doesn't seem to have a way to interface with it without modifying the .py code

**Reply by DigiH** [reply]
As mentioned above, you're looking for a MAC address white-list option, like OpenMQTTGateway has, but Theengs Gateway does not have this option so far. The discovery model_id filter also mentioned above has been implemented to avoid discovery entries for such unwanted devices, but it won't help you currently with your MQTT published messages.

If you have an ESP32 at hand you might want to try [OpenMQTTGateway](https://github.com/1technophile/OpenMQTTGateway) with its white-list filtering if and until this might be implemented in Theengs Gateway.

**Reply by MrSco** [reply]
I’m using raspberry pi’s cause I have so many lying around. I wrote a python script using bleak that sends mqtt signals for the MAC addresses I want only. I guess I’ll just stick with that. 

Thanks for the info. Happy new year!

**Reply by DigiH** [reply]
Feel free to adapt your Python script white-list feature into the Theengs Gateway Python code for a Pull Request, so it might get integrated in one of the next releases.

Thanks, and Happy New Year to you, too!

**Reply by MrSco** [reply]
I think its just a matter of making use of the bleakscanner service_uuids filter parameter in the ble_gateway.py file. 

I'm using tile tags on my cat to control a door I built for him. Tile tags use "0xfeed" for their uuid's. 

So first filter can be service_uuids; on line 354 ...

`if self.configuration["service_uuids"]:
    scanner_kwargs["service_uuids"] = self.configuration["service_uuids"]`

and then second level filter can be on device.address; on line 393:

`if self.configuration["addresses"] and device.address not in self.configuration["addresses"]:
    """Skip Detected device because its not in the list we care about."""
    logger.debug("SKIPPED - %s:%s", device.address, advertisement_data)
    return`

if this makes sense I can do PR but I haven't tested it.

**Reply by DigiH** [reply]
I'm afraid I'll have to leave that discussion to our resident Python wiz, as I've only started myself to submit Python code to Gateway myself recently ;)

The above though seems to be more uuid filtering, which would still also let though any similar devices possibly owned by your neighbours or passer-bys, while a true Bluetooth MAC address filtering would allow for selectively only allowing for desired individual devices of any type to be published.

**Reply by MrSco** [reply]
the two IF statements I posted are for both uuid filtering for only showing Tile tags (0xfeed) in my use case. AND device mac address filtering (AA:BB:CC:DD:EE:FF).  both can be used independently of one another. pretty straight forward, but not sure if using the bluez arguements may interfere

**Comment by koenvervloesem** [comment]
Isn't it more natural to filter on the model ID, like we do for the discovery filter? So don't filter on what the BLE code receives, but filter on what the MQTT code publishes.

**Reply by MrSco** [reply]
sure, my goal was to just not have network traffic about other mac addresses that I want to track.
